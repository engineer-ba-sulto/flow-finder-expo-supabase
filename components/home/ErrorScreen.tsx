import React from "react";
import { ScrollView, Text, View } from "react-native";
import { Button } from "../../components/ui/Button";

interface ErrorScreenProps {
  error: any;
  onRetry: () => void;
}

const ErrorScreen: React.FC<ErrorScreenProps> = ({ error, onRetry }) => (
  <ScrollView
    className="flex-1 bg-white"
    contentContainerStyle={{
      flexGrow: 1,
      justifyContent: "center",
      alignItems: "center",
      paddingHorizontal: 16,
    }}
    accessibilityLabel="エラー画面"
  >
    <View className="items-center">
      <Text
        className="text-xl font-semibold text-red-500 mb-4 text-center"
        accessibilityRole="header"
        accessibilityLabel="エラーが発生しました"
      >
        エラーが発生しました
      </Text>
      <Text
        className="text-center text-gray-600 mb-6 leading-6"
        accessibilityLabel={`エラー詳細: ${error.message}`}
      >
        {error.message}
      </Text>
      <Button
        variant="primary"
        onPress={onRetry}
        accessibilityLabel="再読み込みを実行"
        accessibilityHint="エラーを解決するために再度読み込みを試行します"
      >
        再読み込み
      </Button>
    </View>
  </ScrollView>
);

export default ErrorScreen;
