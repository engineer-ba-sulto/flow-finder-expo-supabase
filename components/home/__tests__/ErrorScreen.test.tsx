import React from 'react';
import { render, screen, fireEvent } from '@testing-library/react-native';
import ErrorScreen from '../ErrorScreen';

// モック設定
jest.mock('../../../components/ui/Button', () => {
  const { Text, TouchableOpacity } = require('react-native');
  return {
    Button: ({ children, onPress, testID, className }: any) => (
      <TouchableOpacity onPress={onPress} testID={testID} className={className}>
        <Text>{children}</Text>
      </TouchableOpacity>
    ),
  };
});

describe('ErrorScreen', () => {
  const defaultProps = {
    error: { message: 'テストエラーメッセージ' },
    onRetry: jest.fn(),
  };

  beforeEach(() => {
    jest.clearAllMocks();
  });

  describe('基本表示', () => {
    it('エラータイトルが表示される', () => {
      render(<ErrorScreen {...defaultProps} />);
      
      expect(screen.getByText('エラーが発生しました')).toBeTruthy();
    });

    it('エラーメッセージが表示される', () => {
      render(<ErrorScreen {...defaultProps} />);
      
      expect(screen.getByText('テストエラーメッセージ')).toBeTruthy();
    });

    it('再読み込みボタンが表示される', () => {
      render(<ErrorScreen {...defaultProps} />);
      
      expect(screen.getByText('再読み込み')).toBeTruthy();
    });
  });

  describe('エラーメッセージのバリエーション', () => {
    it('異なるエラーメッセージが正しく表示される', () => {
      const customProps = {
        ...defaultProps,
        error: { message: '別のエラーメッセージ' },
      };
      
      render(<ErrorScreen {...customProps} />);
      
      expect(screen.getByText('別のエラーメッセージ')).toBeTruthy();
    });

    it('空のエラーメッセージでもレンダリングされる', () => {
      const emptyErrorProps = {
        ...defaultProps,
        error: { message: '' },
      };
      
      render(<ErrorScreen {...emptyErrorProps} />);
      
      expect(screen.getByText('エラーが発生しました')).toBeTruthy();
      expect(screen.getByText('再読み込み')).toBeTruthy();
    });

    it('長いエラーメッセージが正しく表示される', () => {
      const longErrorProps = {
        ...defaultProps,
        error: { 
          message: 'これは非常に長いエラーメッセージです。ネットワークの接続に問題があるか、サーバーが一時的に利用できない状態にある可能性があります。しばらくしてから再度お試しください。'
        },
      };
      
      render(<ErrorScreen {...longErrorProps} />);
      
      expect(screen.getByText(longErrorProps.error.message)).toBeTruthy();
    });
  });

  describe('再読み込み機能', () => {
    it('再読み込みボタンをタップするとonRetryが呼ばれる', () => {
      render(<ErrorScreen {...defaultProps} />);
      
      const retryButton = screen.getByText('再読み込み');
      fireEvent.press(retryButton);
      
      expect(defaultProps.onRetry).toHaveBeenCalledTimes(1);
    });

    it('再読み込みボタンを複数回タップすると複数回onRetryが呼ばれる', () => {
      render(<ErrorScreen {...defaultProps} />);
      
      const retryButton = screen.getByText('再読み込み');
      fireEvent.press(retryButton);
      fireEvent.press(retryButton);
      fireEvent.press(retryButton);
      
      expect(defaultProps.onRetry).toHaveBeenCalledTimes(3);
    });
  });

  describe('レイアウトとスタイル', () => {
    it('ScrollViewが正しく設定されている', () => {
      const { toJSON } = render(<ErrorScreen {...defaultProps} />);
      
      // ScrollViewの構造を確認
      expect(toJSON()).toMatchSnapshot();
    });

    it('中央寄せレイアウトが適用されている', () => {
      render(<ErrorScreen {...defaultProps} />);
      
      // 要素が存在することで中央寄せが機能していることを確認
      expect(screen.getByText('エラーが発生しました')).toBeTruthy();
      expect(screen.getByText('テストエラーメッセージ')).toBeTruthy();
      expect(screen.getByText('再読み込み')).toBeTruthy();
    });
  });

  describe('アクセシビリティ', () => {
    it('適切なアクセシビリティラベルが設定されている', () => {
      render(<ErrorScreen {...defaultProps} />);
      
      expect(screen.getByLabelText('エラー画面')).toBeTruthy();
      expect(screen.getByLabelText('エラーが発生しました')).toBeTruthy();
      expect(screen.getByLabelText('エラー詳細: テストエラーメッセージ')).toBeTruthy();
      expect(screen.getByLabelText('再読み込みを実行')).toBeTruthy();
    });

    it('ヘッダーロールが正しく設定されている', () => {
      const { toJSON } = render(<ErrorScreen {...defaultProps} />);
      
      // accessibilityRoleがheaderに設定されていることを確認
      expect(toJSON()).toMatchSnapshot();
    });
  });

  describe('プロパティの型', () => {
    it('Error オブジェクトの様々な形式に対応する', () => {
      // 標準的なError オブジェクト
      const standardError = new Error('標準エラー');
      const standardProps = {
        ...defaultProps,
        error: standardError,
      };
      
      render(<ErrorScreen {...standardProps} />);
      expect(screen.getByText('標準エラー')).toBeTruthy();
    });

    it('カスタムエラーオブジェクトに対応する', () => {
      const customError = {
        message: 'カスタムエラー',
        code: 'CUSTOM_ERROR',
      };
      const customProps = {
        ...defaultProps,
        error: customError,
      };
      
      render(<ErrorScreen {...customProps} />);
      expect(screen.getByText('カスタムエラー')).toBeTruthy();
    });
  });
});