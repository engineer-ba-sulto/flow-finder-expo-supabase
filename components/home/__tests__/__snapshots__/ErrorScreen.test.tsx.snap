// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ErrorScreen アクセシビリティ ヘッダーロールが正しく設定されている 1`] = `
<RCTScrollView
  accessibilityLabel="エラー画面"
  className="flex-1 bg-white"
  contentContainerStyle={
    {
      "alignItems": "center",
      "flexGrow": 1,
      "justifyContent": "center",
      "paddingHorizontal": 16,
    }
  }
>
  <View>
    <View
      className="items-center"
    >
      <Text
        accessibilityLabel="エラーが発生しました"
        accessibilityRole="header"
        className="text-xl font-semibold text-red-500 mb-4 text-center"
      >
        エラーが発生しました
      </Text>
      <Text
        accessibilityLabel="エラー詳細: テストエラーメッセージ"
        className="text-center text-gray-600 mb-6 leading-6"
      >
        テストエラーメッセージ
      </Text>
      <View
        accessibilityState={
          {
            "busy": undefined,
            "checked": undefined,
            "disabled": undefined,
            "expanded": undefined,
            "selected": undefined,
          }
        }
        accessibilityValue={
          {
            "max": undefined,
            "min": undefined,
            "now": undefined,
            "text": undefined,
          }
        }
        accessible={true}
        collapsable={false}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          {
            "opacity": 1,
          }
        }
      >
        <Text>
          再読み込み
        </Text>
      </View>
    </View>
  </View>
</RCTScrollView>
`;

exports[`ErrorScreen レイアウトとスタイル ScrollViewが正しく設定されている 1`] = `
<RCTScrollView
  accessibilityLabel="エラー画面"
  className="flex-1 bg-white"
  contentContainerStyle={
    {
      "alignItems": "center",
      "flexGrow": 1,
      "justifyContent": "center",
      "paddingHorizontal": 16,
    }
  }
>
  <View>
    <View
      className="items-center"
    >
      <Text
        accessibilityLabel="エラーが発生しました"
        accessibilityRole="header"
        className="text-xl font-semibold text-red-500 mb-4 text-center"
      >
        エラーが発生しました
      </Text>
      <Text
        accessibilityLabel="エラー詳細: テストエラーメッセージ"
        className="text-center text-gray-600 mb-6 leading-6"
      >
        テストエラーメッセージ
      </Text>
      <View
        accessibilityState={
          {
            "busy": undefined,
            "checked": undefined,
            "disabled": undefined,
            "expanded": undefined,
            "selected": undefined,
          }
        }
        accessibilityValue={
          {
            "max": undefined,
            "min": undefined,
            "now": undefined,
            "text": undefined,
          }
        }
        accessible={true}
        collapsable={false}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          {
            "opacity": 1,
          }
        }
      >
        <Text>
          再読み込み
        </Text>
      </View>
    </View>
  </View>
</RCTScrollView>
`;
